File: Common/DelegateEqualityComparer.vb
Namespace Common

    Public Class DelegateEqualityComparer(Of T)
        Inherits EqualityComparer(Of T)

        Public Sub New(ByVal EqualsDelegate As Func(Of T, T, Boolean), ByVal GetHashCodeDelegate As Func(Of T, Integer))
            Me.EqualsDelegate = EqualsDelegate
            Me.GetHashCodeDelegate = GetHashCodeDelegate
        End Sub

        Public Overrides Function Equals(x As T, y As T) As Boolean
            Return Me.EqualsDelegate.Invoke(x, y)
        End Function

        Public Overrides Function GetHashCode(obj As T) As Integer
            Return Me.GetHashCodeDelegate.Invoke(obj)
        End Function

        Private ReadOnly EqualsDelegate As Func(Of T, T, Boolean)
        Private ReadOnly GetHashCodeDelegate As Func(Of T, Integer)

    End Class

End Namespace

File: Common/DelegateEqualityComparer.cs
using System.Collections.Generic;
using System;

namespace Ks.Common
{
    public class DelegateEqualityComparer<T> : EqualityComparer<T>
    {
        public DelegateEqualityComparer(Func<T, T, bool> EqualsDelegate, Func<T, int> GetHashCodeDelegate)
        {
            this.EqualsDelegate = EqualsDelegate;
            this.GetHashCodeDelegate = GetHashCodeDelegate;
        }

        public override bool Equals(T x, T y)
        {
            return this.EqualsDelegate.Invoke(x, y);
        }

        public override int GetHashCode(T obj)
        {
            return this.GetHashCodeDelegate.Invoke(obj);
        }

        private readonly Func<T, T, bool> EqualsDelegate;
        private readonly Func<T, int> GetHashCodeDelegate;
    }
}
